src/file_pipe_log/log_file.rs

```diff
@@ -67,28 +67,26 @@ impl<F: FileSystem> LogFileWriter<F> {
     }
 
     fn write_header(&mut self, format: LogFileFormat) -> IoResult<()> {
-        self.writer.seek(SeekFrom::Start(0))?;
+        self.writer.rewind()?;
         self.written = 0;
         let mut buf = Vec::with_capacity(LogFileFormat::encoded_len(format.version));
         format.encode(&mut buf).unwrap();
         self.write(&buf, 0)
     }
 
-    pub fn close(&mut self) -> IoResult<()> {
+    pub fn close(&mut self) {
```

Comment:
why not let them return the error?

Reply:
Basically, I want the idea of "panic if truncate or sync fail" to be consistent as much as possible cross the whole module. So, it is best to make it as lower level as possible. Then the call sites won't have to debate on the expected behaviors anymore. `truncate` and `sync` do not return errors, so `close` doesn't need to.

Reply:
"truncate" failure is retriable (at least they should). In fact, we rely on it to recover from disk full (after buffered write is full, we still need to remove partial writes from page cache, otherwise in the next recovery we will abandon all writes beyond this point)

Reply:
Truncate is still retriable. For example when closing file before shutdown, its failure is totally okay. The case where we panic when truncate fails is purely for the sake of convenience. We can also retry truncate for a few times. Fsync failure is different, its characteristics is determined by filesystem and we should always panic regardless of the application or engine code context.

Reply:
So, you are proposing that, in this PR. In addition to the change of behavior of `create`, change `truncate` as well?

Reply:
The general guideline is, for errors that can be retried without causing data inconsistency, bubble them out, otherwise panic. However we can choose which layer of interface to enforce this protocol. Previously we enforce it for `PipeLog` trait. By "push down the panic", we need to move this layer to lower modules such as writers and private functions. We can refer to this property as "fail-safe" (which is a lesser guarantee than panic-safe).

To be more specific:
- Make `Writer::sync` fail-safe and panic inside if fsync fails.
- Make `SinglePipe::rotate_imp` fail-safe and panic if `Writer::close` fails, because continuing this this failure would result in data inconsistency at next restart. (This inconsistency is "pipe"-level, not file-level. So we shouldn't panic at file writer level.)
- Bubble any error at `SinglePipe::append`.

Reply:
PTAL